src/systems/filecoin_markets/storage_market/storage_market_actor.go

```diff
@@ -312,6 +319,37 @@ func (a *StorageMarketActorCode_I) GetLastDealExpirationFromDealIDs(rt Runtime,
 	return lastDealExpiration
 }
 
+func (a *StorageMarketActorCode_I) GetUnsealedCIDForDealIDs(rt Runtime, sectorSize util.UVarint, dealIDs []deal.DealID) vmr.InvocOutput {
+	pieceInfos := make([]*sector.PieceInfo_I, len(dealIDs))
+
+	h, st := a.State(rt)
+
+	var totalPieceSize util.UInt
+
+	for index, deal := range st.Deals() {
+		proposal := deal.Proposal()
+		pieceSize := util.UInt(proposal.PieceSize().Total())
+		totalPieceSize += pieceSize
+		pieceInfos[index].PieceCID_ = proposal.PieceCID()
+		pieceInfos[index].Size_ = pieceSize
+	}
+
+	Release(rt, h, st)
+
```

Comment:
@jbenet Got it: this is actually how I had it originally (more or less), but @whyrusleeping noted that passing the `PieceInfo`s through the VM would be difficult and/or expensive from an implementation perspective. I made this change and more convenient proofs API to unblock him. I can change back, but this might be a useful place for you and he to discuss VM pragmatics/philosophy (which will also be valuable for me, since I'm still coming up to speed on the issues). cc: @jzimmerman @zixuanzh

Reply:
- >  more convenient proofs API to unblock him
   
    sgtm. this change says nothing about filproofs API. I think that API should definitely be optimized for callers from anywhere about the specific thing they need.
- This doesn't need to block testnet in implementation. I'm OK with whatever @whyrusleeping wants to do there
- But in the spec, (spec tests, and impl for mainnet) all sector & proof things in the VM need to be scoped to `StorageMinerActor`.  This is an important scope isolation thing.
- We'll need to sort out the larger performance optimization issue. @whyrusleeping is very right to be concerned about this
- This call (returning pieceInfos) is small compared to others, and all of those are small compared to the wasted costs of serializing and writing through during failures, including gas failures. Before mainnet, we'll end up running into these problems and likely optimize things in the VM itself